clamav/cnudie.py

```diff
@@ -21,22 +25,103 @@
 ci.log.configure_default_logging()
 
 
+def _should_skip_scan(
+        resource_node: cnudie.iter.ResourceNode,
+        current_version_info: clamav.model.ClamAVVersionInfo,
+        virus_db_max_age_days: int,
+        skip_if_virus_definition_unchanged: bool,
+        delivery_client: delivery.client.DeliveryServiceClient = None,
+) -> bool:
+
+    if not delivery_client:
+        return False
+
+    previous_findings = delivery_client.artefact_metadata_for_resource_node(
+        resource_node=resource_node,
+        types=[dso.model.Datatype.MALWARE],
+    )
+
+    sorted_findings = sorted(
+        previous_findings,
+        key=lambda f: f.meta.creation_date.astimezone(),
+        reverse=True,
+    )
+
+    if sorted_findings:
+        latest_finding = sorted_findings[0]
+    else:
+        return False
+
+    if isinstance(latest_finding.data, dict):
+        # In this case the previous scan we have available predates the current dataclasses.
+        # Returning False here results in a rescan which will put a new scan result into
+        # the database if there are any findings.
+        # Also, as an additional safeguard against future errors, we check that no such
+        # supposedly structurally outdated finding was put into the database after this
+        # change was added.
+        cutoff_date = datetime.datetime(year=2023, month=4, day=1)
+        if latest_finding.meta.creation_date > cutoff_date:
+            raise RuntimeError(
+                'Encountered a malware scan result with outdated schema that was created '
+                'after introduction of the new schema.'
+            )
+        return False
+
+    finding_clamav_metadata = latest_finding.data.metadata
+
+    if (
+        skip_if_virus_definition_unchanged
```

Comment:
maybe we already discussed this: _why_ would we want to rescan upon unchanged virus definitions?